### ISSUE TITLE
RWX azurefiles contention

### ISSUE THREAD

**Issue by marto** [Azure/AKS#1232]
We using `kubernetes.io/azure-file` on a AKS v14.15.5 cluster and having issues around reading files when there is a bit of contention on the file system.

We using a shared `RWX` storage to act as a crude file based shared cache between a small number of containers (less than 4) in a cluster. We have setup a custom storage class and are using it in a stateful set. No issues there, it mounts, when one server adds a file, the file also appears in the other container.

Under the hood, the software creates a temporary file in the dir the cached file will ultimately reside in, writes the contents to the temp file, closes it and then performs a move operation to rename the file to it's intended name. We expect there to be a chance of a race between two or more processes (in the same container or on another container) trying to write to the cache and perform the final file move operation almost at the same time. We expect one of the move operations to fail, and don't care if the other file move operation fails as long as the file eventually become consistent across all nodes and as long as the file can be read (i.e. we ok with dirty reads for a while).

**What happened**:
Under a small load the destination file become unreadable and although the file exists on the file system (you can see it via `ls`) it can't be read and we get a `File not found error` when performing any operation on the file. We can't seem to be able to recover from this, can't do any operations on the file (even can't remove it) and have to scale down the cluster to 0 replicas and scale it back up (i.e. force a remount) to recover from this situation so we can finally read the file contents.

**What you expected to happen**:
We expect that if the file exists on the file system it should always be in a state where we can read the contents of the file. It's acceptable for us to read the contents of the file and get `dirty reads` i.e. a local cached version of the file is available rather than the new fresh version.

**How to reproduce it (as minimally and precisely as possible)**:
```
## 00-aks-storage-classes.yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: cache-data
provisioner: kubernetes.io/azure-file
mountOptions:
  - dir_mode=0775
  - file_mode=0660
  - uid=0033
  - gid=0033
parameters:
  skuName: Standard_LRS
```

Setup up a stateful set with a dynamic volume PV claim using the above `cache-data` storage class and mount it on `/sharedfs`. Get each container to run a script that creates a bit of contention of the shared fs; for example:
```
#!/bin/bash
random() {
  expr $(od -A n -t d -N 1 /dev/urandom) % 5
}

mount=/sharedfs
while [ true ]; do
    dir="$mount/foo/$(random)/bar/"
    file="$dir/foo"
    mkdir -p "$dir"
    dd if=/dev/urandom count=100 bs=1024 of="${file}.$$"
    mv "${file}.$$" "${file}.data"
    cat "${file}.data" | md5sum
    sleep 0.0$(random)
done
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-18T14:36:53Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.5", GitCommit:"0e9fcb426b100a2aea5ed5c25b3d8cfbb01a8acf", GitTreeState:"clean", BuildDate:"2019-08-05T09:13:08Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Size of cluster (how many worker nodes are in the cluster?)
5
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.)
Apache HTTPD with a shared FS cache
- Others:

**Comment by IvoZheglov** [comment]
This is an issue on the Linux CIFS side. See below:

https://patchwork.ozlabs.org/patch/1169516/
https://patchwork.ozlabs.org/patch/1169517/
https://patchwork.ozlabs.org/patch/1169518/

**Comment by palma21** [comment]
It does look related, and all those commits should be on the now available AKS Ubuntu 18.04 base image. Closing based on that, but feel free to reply if you believe it's not correct and I'm happy to re-open.
